### ISSUE TITLE
Remove duplication from CI Release workflow

### ISSUE THREAD

**Issue by gakonst** [ethereum-optimism/optimism#482]
# The Problem

The release code is [heavily duplicated](https://github.com/ethereum-optimism/optimism/blob/master/.github/workflows/release.yml#L64-L202), in order to allow for dynamically launching jobs, conditional on the job being output by `changesets`.

The main challenge is basically having conditional/dynamic logic in the matrix generation, so that the jobs generated only match the published packages (whereas now we have hard-coded the jobs and conditionally trigger them)

# The Solution

Ideally, what we'd want, is a dynamic matric using `publishedPackages` which builds/pushes the services, and also "intelligently" builds the `builder` container only once (instead of having to re-build it for each service).

# Another solution

Maybe we do not need to intelligently build the builder image, and just re-building it once per service is fine.

**Comment by platocrat** [comment]
> # Another solution
> Maybe we do not need to intelligently build the builder image, and just re-building it once per service is fine.

This is an adjacent issue, but occasionally running the docker services after having built and run them results in an error returned by docker.
@TransmissionsDev and I experienced this Friday and Saturday, and today.

**Comment by gakonst** [comment]
@platocrat please open a separate issue with the steps to reproduce (add a note if they need to be done "some" times to get the error) and the logs returned by docker!
